File: examples/rust/examples/client_websocket_subscribe.rs
//! WebSocket Subscription Example - Real-time mutation notifications
//!
//! Demonstrates subscribing to collection changes via WebSocket and
//! receiving push notifications when records are inserted, updated, or deleted.
//!
//! Requires: ekoDB server running on localhost:8080

use futures_util::{SinkExt, StreamExt};
use reqwest::StatusCode;
use serde_json::{json, Value};
use std::{env, error::Error, time::Duration};
use tokio_tungstenite::{connect_async, tungstenite::protocol::Message};
use url::Url;

async fn delete_collection(
    client: &reqwest::Client,
    base_url: &str,
    token: &str,
    collection: &str,
) -> Result<bool, Box<dyn Error>> {
    let response = client
        .delete(format!("{base_url}/api/collections/{collection}"))
        .bearer_auth(token)
        .send()
        .await?;
    if response.status() == StatusCode::NOT_FOUND {
        return Ok(false);
    }
    response.error_for_status()?;
    Ok(true)
}

async fn next_json<S>(socket: &mut S, timeout: Duration) -> Result<Value, Box<dyn Error>>
where
    S: futures_util::Stream<Item = Result<Message, tokio_tungstenite::tungstenite::Error>>
        + futures_util::Sink<Message, Error = tokio_tungstenite::tungstenite::Error>
        + Unpin,
{
    tokio::time::timeout(timeout, async {
        loop {
            let message = socket
                .next()
                .await
                .ok_or("WebSocket closed before the expected message")??;
            match message {
                Message::Text(text) => return Ok(serde_json::from_str(&text)?),
                Message::Ping(payload) => socket.send(Message::Pong(payload)).await?,
                Message::Pong(_) | Message::Frame(_) => {}
                Message::Close(frame) => {
                    return Err(
                        format!("WebSocket closed before the expected message: {frame:?}").into(),
                    );
                }
                Message::Binary(_) => {
                    return Err("expected a text WebSocket message, got binary data".into());
                }
            }
        }
    })
    .await
    .map_err(|_| "timed out waiting for WebSocket message")?
}

fn required_id<'a>(response: &'a Value, operation: &str) -> Result<&'a str, Box<dyn Error>> {
    response["id"]
        .as_str()
        .filter(|id| !id.is_empty())
        .ok_or_else(|| format!("{operation} did not return a non-empty record ID").into())
}

fn validate_notification<'a>(
    notification: &'a Value,
    collection: &str,
    record_id: &str,
) -> Result<&'a Value, Box<dyn Error>> {
    if notification["type"] != "MutationNotification" {
        return Err(format!("unexpected WebSocket message: {notification}").into());
    }
    let payload = notification["payload"]
        .as_object()
        .ok_or("mutation notification did not include an object payload")?;
    if payload.get("event").and_then(Value::as_str) != Some("insert") {
        return Err(format!("expected insert notification: {notification}").into());
    }
    if payload.get("collection").and_then(Value::as_str) != Some(collection) {
        return Err(format!("notification targeted the wrong collection: {notification}").into());
    }
    let record_ids = payload
        .get("record_ids")
        .and_then(Value::as_array)
        .ok_or("mutation notification did not include record_ids")?;
    if !record_ids.iter().any(|id| id.as_str() == Some(record_id)) {
        return Err(format!("notification did not include inserted ID {record_id}").into());
    }
    if payload.get("timestamp").and_then(Value::as_str).is_none() {
        return Err("mutation notification did not include a timestamp".into());
    }
    Ok(&notification["payload"])
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    dotenv::dotenv().ok();

    let base_url = env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string());
    let ws_url = env::var("WS_BASE_URL").unwrap_or_else(|_| "ws://localhost:8080".to_string());
    let api_key = env::var("API_BASE_KEY").unwrap_or_else(|_| "a-test-api-key-from-ekodb".into());
    let collection = "ws_subscribe_example_rs";

    // Step 1: Authenticate
    let http_client = reqwest::Client::new();
    let token_data: Value = http_client
        .post(format!("{base_url}/api/auth/token"))
        .json(&json!({ "api_key": api_key }))
        .send()
        .await?
        .error_for_status()?
        .json()
        .await?;
    let token = token_data["token"]
        .as_str()
        .filter(|token| !token.is_empty())
        .ok_or("authentication response did not include a token")?
        .to_string();
    println!("✓ Authentication successful");
    delete_collection(&http_client, &base_url, &token, collection).await?;

    let operation_result: Result<(), Box<dyn Error>> = async {
        // Step 2: Connect to WebSocket
        println!("\n=== Connecting to WebSocket ===");
        let url = format!("{ws_url}/api/ws");
        let parsed = Url::parse(&url)?;
        let host = parsed
            .host_str()
            .map(|host| match parsed.port() {
                Some(port) => format!("{host}:{port}"),
                None => host.to_string(),
            })
            .unwrap_or_else(|| "localhost:8080".to_string());
        let request = tokio_tungstenite::tungstenite::http::Request::builder()
            .uri(&url)
            .header("Authorization", format!("Bearer {token}"))
            .header("Host", &host)
            .header("Connection", "Upgrade")
            .header("Upgrade", "websocket")
            .header("Sec-WebSocket-Version", "13")
            .header(
                "Sec-WebSocket-Key",
                tokio_tungstenite::tungstenite::handshake::client::generate_key(),
            )
            .body(())?;
        let (mut socket, _) = connect_async(request).await?;
        println!("✓ WebSocket connected");

        // Step 3: Subscribe to collection
        println!("\n=== Subscribing to '{collection}' ===");
        let subscribe_msg = json!({
            "type": "Subscribe",
            "payload": { "collection": collection }
        });
        socket
            .send(Message::Text(subscribe_msg.to_string().into()))
            .await?;
        let response = next_json(&mut socket, Duration::from_secs(10)).await?;
        let sub_id = response["payload"]["data"]["subscription_id"]
            .as_str()
            .filter(|id| !id.is_empty())
            .ok_or_else(|| format!("invalid subscription response: {response}"))?;
        println!("✓ Subscribed (subscription_id: {sub_id})");

        // Step 4: Insert records via REST API to trigger notifications
        println!("\n=== Performing mutations to trigger notifications ===");
        println!("Inserting record 1...");
        let insert_data: Value = http_client
            .post(format!("{base_url}/api/insert/{collection}"))
            .bearer_auth(&token)
            .json(&json!({ "name": "Alice", "role": "engineer", "active": true }))
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;
        let record1_id = required_id(&insert_data, "first insert")?;
        println!("✓ Inserted: {record1_id}");
        let notification = match next_json(&mut socket, Duration::from_secs(5)).await {
            Ok(notification) => notification,
            Err(error) => {
                if error.to_string().contains("timed out") {
                    println!("  ⏳ No notification within timeout");
                }
                return Err(error);
            }
        };
        let payload = validate_notification(&notification, collection, record1_id)?;
        println!("\n  📡 Notification received:");
        println!("     Event:      {}", payload["event"]);
        println!("     Collection: {}", payload["collection"]);
        println!("     Record IDs: {}", payload["record_ids"]);
        println!("     Timestamp:  {}", payload["timestamp"]);

        println!("\nInserting record 2...");
        let insert_data2: Value = http_client
            .post(format!("{base_url}/api/insert/{collection}"))
            .bearer_auth(&token)
            .json(&json!({ "name": "Bob", "role": "designer", "active": true }))
            .send()
            .await?
            .error_for_status()?
            .json()
            .await?;
        let record2_id = required_id(&insert_data2, "second insert")?;
        println!("✓ Inserted: {record2_id}");
        let notification = match next_json(&mut socket, Duration::from_secs(5)).await {
            Ok(notification) => notification,
            Err(error) => {
                if error.to_string().contains("timed out") {
                    println!("  ⏳ No notification within timeout");
                }
                return Err(error);
            }
        };
        let payload = validate_notification(&notification, collection, record2_id)?;
        println!("\n  📡 Notification received:");
        println!("     Event:      {}", payload["event"]);
        println!("     Record IDs: {}", payload["record_ids"]);

        // Step 5: Unsubscribe
        println!("\n=== Unsubscribing ===");
        let unsub_msg = json!({
            "type": "Unsubscribe",
            "payload": { "collection": collection }
        });
        socket
            .send(Message::Text(unsub_msg.to_string().into()))
            .await?;
        let unsub_response = next_json(&mut socket, Duration::from_secs(10)).await?;
        let unsub_data = &unsub_response["payload"]["data"];
        if unsub_data["collection"] != collection
            || unsub_data["found"] != true
            || unsub_data["unsubscribed"] != true
        {
            return Err(format!("invalid unsubscribe response: {unsub_response}").into());
        }
        println!("✓ Unsubscribed");
        Ok(())
    }
    .await;

    println!("\n=== Cleanup ===");
    let cleanup_result = delete_collection(&http_client, &base_url, &token, collection).await;
    if matches!(cleanup_result, Ok(true)) {
        println!("✓ Deleted collection '{collection}'");
    }
    match (operation_result, cleanup_result) {
        (Err(primary), Err(cleanup)) => {
            return Err(format!("{primary}; cleanup also failed: {cleanup}").into());
        }
        (Err(primary), _) => return Err(primary),
        (Ok(()), Err(cleanup)) => return Err(cleanup),
        (Ok(()), Ok(_)) => {}
    }

    println!("\n✓ WebSocket subscription example completed successfully");
    Ok(())
}

File: examples/kotlin/examples/ClientWebsocketSubscribe.kt
package io.ekodb.client.examples

import io.github.cdimascio.dotenv.dotenv
import io.ktor.client.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.contentnegotiation.*
import io.ktor.client.plugins.websocket.*
import io.ktor.client.request.*
import io.ktor.client.statement.*
import io.ktor.http.*
import io.ktor.serialization.kotlinx.json.*
import io.ktor.websocket.*
import kotlinx.coroutines.*
import kotlinx.coroutines.channels.Channel
import kotlinx.serialization.json.*

/**
 * WebSocket Subscription Example - Real-time mutation notifications
 *
 * Demonstrates subscribing to collection changes via WebSocket and
 * receiving push notifications when records are inserted, updated, or deleted.
 *
 * Requires: ekoDB server running on localhost:8080
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val wsUrl = dotenv["WS_BASE_URL"] ?: "ws://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val collection = "ws_subscribe_example_kt"

    println("=== WebSocket Subscription Example (Kotlin) ===\n")

    val httpClient = HttpClient(CIO) {
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true })
        }
        install(WebSockets)
    }

    var token: String? = null
    var runFailure: Throwable? = null

    try {
        // Step 1: Authenticate
        val tokenResponse = httpClient.post("$baseUrl/api/auth/token") {
            contentType(ContentType.Application.Json)
            setBody(buildJsonObject { put("api_key", apiKey) }.toString())
        }
        val tokenJson = Json.parseToJsonElement(tokenResponse.bodyAsText()).jsonObject
        token = tokenJson["token"]!!.jsonPrimitive.content
        println("✓ Authentication successful")

        val staleCleanup = httpClient.delete("$baseUrl/api/collections/$collection") {
            header("Authorization", "Bearer $token")
        }
        if (!staleCleanup.status.isSuccess() && staleCleanup.status != HttpStatusCode.NotFound) {
            error("Stale collection cleanup failed: ${staleCleanup.status} ${staleCleanup.bodyAsText()}")
        }

        // Step 2: Connect to WebSocket
        println("\n=== Connecting to WebSocket ===")

        val wsUrlParsed = Url(wsUrl)
        val notifications = Channel<JsonObject>(10)
        val responses = Channel<JsonObject>(10)

        httpClient.webSocket(
            method = HttpMethod.Get,
            host = wsUrlParsed.host,
            port = wsUrlParsed.port,
            path = "/api/ws",
            request = {
                header("Authorization", "Bearer $token")
            }
        ) {
            println("✓ WebSocket connected")

            // Step 3: Subscribe to collection
            println("\n=== Subscribing to '$collection' ===")

            val subscribeMsg = buildJsonObject {
                put("type", "Subscribe")
                put("payload", buildJsonObject {
                    put("collection", collection)
                })
            }
            send(Frame.Text(subscribeMsg.toString()))

            // Read subscription confirmation
            val confirmFrame = withTimeout(5000) { incoming.receive() }
            require(confirmFrame is Frame.Text) { "Subscription confirmation was not text" }
            val response = Json.parseToJsonElement(confirmFrame.readText()).jsonObject
            require(response["type"]?.jsonPrimitive?.content == "Success") {
                "Subscription failed: $response"
            }
            val subId = response["payload"]
                ?.jsonObject?.get("data")
                ?.jsonObject?.get("subscription_id")
                ?.jsonPrimitive?.content
                ?: error("Subscription confirmation omitted subscription_id: $response")
            println("✓ Subscribed (subscription_id: $subId)")

            // Step 4: Launch a coroutine to listen for notifications
            val readJob = launch {
                for (frame in incoming) {
                    if (frame is Frame.Text) {
                        val parsed = Json.parseToJsonElement(frame.readText()).jsonObject
                        if (parsed["type"]?.jsonPrimitive?.content == "MutationNotification") {
                            notifications.send(parsed)
                        } else {
                            responses.send(parsed)
                        }
                    }
                }
            }

            // Step 5: Insert records via REST API to trigger notifications
            println("\n=== Performing mutations to trigger notifications ===")

            println("Inserting record 1...")
            val insertRes1 = httpClient.post("$baseUrl/api/insert/$collection") {
                header("Authorization", "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(buildJsonObject {
                    put("name", "Alice")
                    put("role", "engineer")
                    put("active", JsonPrimitive(true))
                }.toString())
            }
            require(insertRes1.status.isSuccess()) {
                "First insert failed: ${insertRes1.status} ${insertRes1.bodyAsText()}"
            }
            val insertData1 = Json.parseToJsonElement(insertRes1.bodyAsText()).jsonObject
            println("✓ Inserted: ${insertData1["id"]}")

            // Wait for notification
            val notification1 = withTimeout(5000) { notifications.receive() }
            val payload1 = notification1["payload"]!!.jsonObject
            require(payload1["event"]?.jsonPrimitive?.content == "insert") { "Unexpected event: $payload1" }
            require(payload1["collection"]?.jsonPrimitive?.content == collection) { "Unexpected collection: $payload1" }
            require(payload1["record_ids"]?.jsonArray?.any { it.jsonPrimitive.content == insertData1["id"]?.jsonPrimitive?.content } == true) {
                "First notification omitted inserted record: $payload1"
            }
            println("\n  📡 Notification received:")
            println("     Event:      ${payload1["event"]}")
            println("     Collection: ${payload1["collection"]}")
            println("     Record IDs: ${payload1["record_ids"]}")
            println("     Timestamp:  ${payload1["timestamp"]}")

            println("\nInserting record 2...")
            val insertRes2 = httpClient.post("$baseUrl/api/insert/$collection") {
                header("Authorization", "Bearer $token")
                contentType(ContentType.Application.Json)
                setBody(buildJsonObject {
                    put("name", "Bob")
                    put("role", "designer")
                    put("active", JsonPrimitive(true))
                }.toString())
            }
            require(insertRes2.status.isSuccess()) {
                "Second insert failed: ${insertRes2.status} ${insertRes2.bodyAsText()}"
            }
            val insertData2 = Json.parseToJsonElement(insertRes2.bodyAsText()).jsonObject
            println("✓ Inserted: ${insertData2["id"]}")

            val notification2 = withTimeout(5000) { notifications.receive() }
            val payload2 = notification2["payload"]!!.jsonObject
            require(payload2["event"]?.jsonPrimitive?.content == "insert") { "Unexpected event: $payload2" }
            require(payload2["collection"]?.jsonPrimitive?.content == collection) { "Unexpected collection: $payload2" }
            require(payload2["record_ids"]?.jsonArray?.any { it.jsonPrimitive.content == insertData2["id"]?.jsonPrimitive?.content } == true) {
                "Second notification omitted inserted record: $payload2"
            }
            println("\n  📡 Notification received:")
            println("     Event:      ${payload2["event"]}")
            println("     Record IDs: ${payload2["record_ids"]}")

            // Step 6: Unsubscribe
            println("\n=== Unsubscribing ===")
            val unsubMsg = buildJsonObject {
                put("type", "Unsubscribe")
                put("payload", buildJsonObject {
                    put("collection", collection)
                })
            }
            send(Frame.Text(unsubMsg.toString()))
            val unsubResponse = withTimeout(5000) { responses.receive() }
            require(unsubResponse["type"]?.jsonPrimitive?.content == "Success") {
                "Unsubscribe failed: $unsubResponse"
            }
            println("✓ Unsubscribed: $unsubResponse")

            // Cleanup
            readJob.cancelAndJoin()
            println("\n✓ WebSocket subscription example completed successfully")
        }
    } catch (error: Throwable) {
        runFailure = error
        throw error
    } finally {
        try {
            token?.let {
                val cleanup = httpClient.delete("$baseUrl/api/collections/$collection") {
                    header("Authorization", "Bearer $it")
                }
                if (!cleanup.status.isSuccess() && cleanup.status != HttpStatusCode.NotFound) {
                    error("Collection cleanup failed: ${cleanup.status} ${cleanup.bodyAsText()}")
                }
                println("✓ Deleted collection '$collection'")
            }
        } catch (cleanupError: Throwable) {
            if (runFailure != null) runFailure.addSuppressed(cleanupError)
            else throw cleanupError
        } finally {
            httpClient.close()
        }
    }
}
